## Pandas deep dive
Lets review some of the key concepts for Pandas and Numpy that we will use during this course. First lets import our libraries, Pandas. 

In [1]:
import pandas as pd

A Pandas Series is a unidimentional matrix of indexed data. We can create one from a list, like this example

In [2]:
data = pd.Series([0.25, 0.5, 0.75, 1.0])
data

0    0.25
1    0.50
2    0.75
3    1.00
dtype: float64

We can see in our output that the Series is wrpaped by a sequence of values and a sequence of indexes. Values are simply a Numpy matrix

In [3]:
data.values

array([0.25, 0.5 , 0.75, 1.  ])

And the index is a matrix of type pd.Index

In [4]:
data.index

RangeIndex(start=0, stop=4, step=1)

We can access data through the associated index

In [5]:
data[1]

0.5

And do data slicing

In [6]:
data[1:3]

1    0.50
2    0.75
dtype: float64

Now, the main difference between Numpy indices and the Series object is that the Series index can be something other than an integer, we can use strings for our index.

In [7]:
data = pd.Series([0.25, 0.5, 0.75, 1.0],
                 index=['a', 'b', 'c', 'd'])
data

a    0.25
b    0.50
c    0.75
d    1.00
dtype: float64

In [8]:
data['b']

0.5

We can think of the Series as a specialized dictionary. We can even create a Pandas Series object from a Python dictionary

In [9]:
mass_dict = {'Sun': "1.989 × 10^30 kg",
                   'Mercury': "3.285 × 10^23 kg",
                   'Venus': "4.867 × 10^24 kg",
                   'Earth': "5.972 × 10^24 kg",
                   'Mars': "6.39 × 10^23 kg"}
mass = pd.Series(mass_dict)
mass

Sun        1.989 × 10^30 kg
Mercury    3.285 × 10^23 kg
Venus      4.867 × 10^24 kg
Earth      5.972 × 10^24 kg
Mars        6.39 × 10^23 kg
dtype: object

We can slice our Series

In [10]:
mass['Mercury':'Earth']

Mercury    3.285 × 10^23 kg
Venus      4.867 × 10^24 kg
Earth      5.972 × 10^24 kg
dtype: object

* The pandas Dataframe*
The next key object in Pandas is the DataFrame. This object can be considered a generalization of a matrix.

This object can be thinked of an ordered sequence of columns, sharing a row index. Lets create a new Series and then use this Series to create a Dataframe

In [11]:
grav_dict = {'Sun': "274 m/s²", 'Mercury': "3.7 m/s²", 'Venus': "8.87 m/s²",
             'Earth': "9.807 m/s²", 'Mars': "3.721 m/s²"}
grav = pd.Series(grav_dict)
grav

Sun          274 m/s²
Mercury      3.7 m/s²
Venus       8.87 m/s²
Earth      9.807 m/s²
Mars       3.721 m/s²
dtype: object

In [12]:
# lets use our mass dictionary from previous explanation
# Create a single Dataframe from both
objects = pd.DataFrame({'mass': mass,
                       'grav': grav})
objects

mass        grav
Sun      1.989 × 10^30 kg    274 m/s²
Mercury  3.285 × 10^23 kg    3.7 m/s²
Venus    4.867 × 10^24 kg   8.87 m/s²
Earth    5.972 × 10^24 kg  9.807 m/s²
Mars      6.39 × 10^23 kg  3.721 m/s²

We can acces each object with its index

In [13]:
objects.index

Index(['Sun', 'Mercury', 'Venus', 'Earth', 'Mars'], dtype='object')

In [14]:
# read the Dataframe columns
objects.columns

Index(['mass', 'grav'], dtype='object')

In [15]:
# Access one of the columns, similar to a dictionary
objects['grav']

Sun          274 m/s²
Mercury      3.7 m/s²
Venus       8.87 m/s²
Earth      9.807 m/s²
Mars       3.721 m/s²
Name: grav, dtype: object

Please notice that we are calling the Dataframe *column*

In [16]:
objects['mass']

Sun        1.989 × 10^30 kg
Mercury    3.285 × 10^23 kg
Venus      4.867 × 10^24 kg
Earth      5.972 × 10^24 kg
Mars        6.39 × 10^23 kg
Name: mass, dtype: object

You can describe a Pandas Dataframe with `.describe()`

In [17]:
objects.describe()

mass      grav
count                  5         5
unique                 5         5
top     1.989 × 10^30 kg  274 m/s²
freq                   1         1

### Optional Exercise
- Create a new dataframe from the following data
`data = {'ages': [23, 24, 25, 26, 27, 28], 'weight': [120, 130, 150, 160, 180, 190]}
- You should now have two data columns `ages` and `weight`. Print only the `ages` column.
- Select the row with age = 27 using `.loc[]`

Lets open our cereal csv and check some Pandas functions.
You can use `head()` to show the first 5 rows. 
You can use `tail()` to show the last 5 rows.

In [18]:
cereal = pd.read_csv('cereal.csv', index_col='name')
cereal.head()

mfr type  calories  protein  fat  sodium  fiber  \
name                                                                        
100% Bran                   N    C        70        4    1     130   10.0   
100% Natural Bran           Q    C       120        3    5      15    2.0   
All-Bran                    K    C        70        4    1     260    9.0   
All-Bran with Extra Fiber   K    C        50        4    0     140   14.0   
Almond Delight              R    C       110        2    2     200    1.0   

                           carbo  sugars  potass  vitamins  shelf  weight  \
name                                                                        
100% Bran                    5.0       6     280        25      3     1.0   
100% Natural Bran            8.0       8     135         0      3     1.0   
All-Bran                     7.0       5     320        25      3     1.0   
All-Bran with Extra Fiber    8.0       0     330        25      3     1.0   
Almond Delight              14.0       8      -1        25      3     1.0   

                           cups     rating  
name                                        
100% Bran                  0.33  68.402973  
100% Natural Bran          1.00  33.983679  
All-Bran                   0.33  59.425505  
All-Bran with Extra Fiber  0.50  93.704912  
Almond Delight             0.75  34.384843

In [19]:
cereal.tail()

mfr type  calories  protein  fat  sodium  fiber  carbo  \
name                                                                         
Triples               G    C       110        2    1     250    0.0   21.0   
Trix                  G    C       110        1    1     140    0.0   13.0   
Wheat Chex            R    C       100        3    1     230    3.0   17.0   
Wheaties              G    C       100        3    1     200    3.0   17.0   
Wheaties Honey Gold   G    C       110        2    1     200    1.0   16.0   

                     sugars  potass  vitamins  shelf  weight  cups     rating  
name                                                                           
Triples                   3      60        25      3     1.0  0.75  39.106174  
Trix                     12      25        25      2     1.0  1.00  27.753301  
Wheat Chex                3     115        25      1     1.0  0.67  49.787445  
Wheaties                  3     110        25      1     1.0  1.00  51.592193  
Wheaties Honey Gold       8      60        25      1     1.0  0.75  36.187559

In [20]:
cereal.describe()

calories    protein        fat      sodium      fiber      carbo  \
count   77.000000  77.000000  77.000000   77.000000  77.000000  77.000000   
mean   106.883117   2.545455   1.012987  159.675325   2.151948  14.597403   
std     19.484119   1.094790   1.006473   83.832295   2.383364   4.278956   
min     50.000000   1.000000   0.000000    0.000000   0.000000  -1.000000   
25%    100.000000   2.000000   0.000000  130.000000   1.000000  12.000000   
50%    110.000000   3.000000   1.000000  180.000000   2.000000  14.000000   
75%    110.000000   3.000000   2.000000  210.000000   3.000000  17.000000   
max    160.000000   6.000000   5.000000  320.000000  14.000000  23.000000   

          sugars      potass    vitamins      shelf     weight       cups  \
count  77.000000   77.000000   77.000000  77.000000  77.000000  77.000000   
mean    6.922078   96.077922   28.246753   2.207792   1.029610   0.821039   
std     4.444885   71.286813   22.342523   0.832524   0.150477   0.232716   
min    -1.000000   -1.000000    0.000000   1.000000   0.500000   0.250000   
25%     3.000000   40.000000   25.000000   1.000000   1.000000   0.670000   
50%     7.000000   90.000000   25.000000   2.000000   1.000000   0.750000   
75%    11.000000  120.000000   25.000000   3.000000   1.000000   1.000000   
max    15.000000  330.000000  100.000000   3.000000   1.500000   1.500000   

          rating  
count  77.000000  
mean   42.665705  
std    14.047289  
min    18.042851  
25%    33.174094  
50%    40.400208  
75%    50.828392  
max    93.704912

In [21]:
cereal.columns

Index(['mfr', 'type', 'calories', 'protein', 'fat', 'sodium', 'fiber', 'carbo',
       'sugars', 'potass', 'vitamins', 'shelf', 'weight', 'cups', 'rating'],
      dtype='object')

In [22]:
# Returns an array of the unique manufacturers
cereal.mfr.unique()

array(['N', 'Q', 'K', 'R', 'G', 'P', 'A'], dtype=object)

You can get info about the dataset by using the `info()` function, giving you the number of rows and columns, rows with non null values, type of data in each column and memory use.

In [23]:
cereal.info()

<class 'pandas.core.frame.DataFrame'>
Index: 77 entries, 100% Bran to Wheaties Honey Gold
Data columns (total 15 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   mfr       77 non-null     object 
 1   type      77 non-null     object 
 2   calories  77 non-null     int64  
 3   protein   77 non-null     int64  
 4   fat       77 non-null     int64  
 5   sodium    77 non-null     int64  
 6   fiber     77 non-null     float64
 7   carbo     77 non-null     float64
 8   sugars    77 non-null     int64  
 9   potass    77 non-null     int64  
 10  vitamins  77 non-null     int64  
 11  shelf     77 non-null     int64  
 12  weight    77 non-null     float64
 13  cups      77 non-null     float64
 14  rating    77 non-null     float64
dtypes: float64(5), int64(8), object(2)
memory usage: 9.6+ KB


You can also output the shape of the dataframe with `shape`

In [24]:
#Should output 77 rows and 15 columns
cereal.shape

(77, 15)

You can use `.loc` to access specific data.
Lets return the cereals which have a protein content higher that 4.0

In [25]:
cereal.loc[cereal['protein'] >= 4.0]

mfr type  calories  protein  fat  sodium  \
name                                                                         
100% Bran                           N    C        70        4    1     130   
All-Bran                            K    C        70        4    1     260   
All-Bran with Extra Fiber           K    C        50        4    0     140   
Cheerios                            G    C       110        6    2     290   
Life                                Q    C       100        4    2     150   
Maypo                               A    H       100        4    1       0   
Muesli Raisins; Dates; & Almonds    R    C       150        4    3      95   
Muesli Raisins; Peaches; & Pecans   R    C       150        4    3     150   
Quaker Oat Squares                  Q    C       100        4    1     135   
Quaker Oatmeal                      Q    H       100        5    2       0   
Special K                           K    C       110        6    0     230   

                                   fiber  carbo  sugars  potass  vitamins  \
name                                                                        
100% Bran                           10.0    5.0       6     280        25   
All-Bran                             9.0    7.0       5     320        25   
All-Bran with Extra Fiber           14.0    8.0       0     330        25   
Cheerios                             2.0   17.0       1     105        25   
Life                                 2.0   12.0       6      95        25   
Maypo                                0.0   16.0       3      95        25   
Muesli Raisins; Dates; & Almonds     3.0   16.0      11     170        25   
Muesli Raisins; Peaches; & Pecans    3.0   16.0      11     170        25   
Quaker Oat Squares                   2.0   14.0       6     110        25   
Quaker Oatmeal                       2.7   -1.0      -1     110         0   
Special K                            1.0   16.0       3      55        25   

                                   shelf  weight  cups     rating  
name                                                               
100% Bran                              3     1.0  0.33  68.402973  
All-Bran                               3     1.0  0.33  59.425505  
All-Bran with Extra Fiber              3     1.0  0.50  93.704912  
Cheerios                               1     1.0  1.25  50.764999  
Life                                   2     1.0  0.67  45.328074  
Maypo                                  2     1.0  1.00  54.850917  
Muesli Raisins; Dates; & Almonds       3     1.0  1.00  37.136863  
Muesli Raisins; Peaches; & Pecans      3     1.0  1.00  34.139765  
Quaker Oat Squares                     3     1.0  0.50  49.511874  
Quaker Oatmeal                         1     1.0  0.67  50.828392  
Special K                              1     1.0  1.00  53.131324

Return cereals with protein higher than 2 grams and sugar lower than 6 grams

In [26]:
cereal.loc[(cereal['protein'] >= 2) & (cereal['sugars'] <= 6)]

mfr type  calories  protein  fat  sodium  fiber  \
name                                                                          
100% Bran                     N    C        70        4    1     130   10.0   
All-Bran                      K    C        70        4    1     260    9.0   
All-Bran with Extra Fiber     K    C        50        4    0     140   14.0   
Bran Chex                     R    C        90        2    1     200    4.0   
Bran Flakes                   P    C        90        3    0     210    5.0   
Cheerios                      G    C       110        6    2     290    2.0   
Corn Chex                     R    C       110        2    0     280    0.0   
Corn Flakes                   K    C       100        2    0     290    1.0   
Cream of Wheat (Quick)        N    H       100        3    0      80    1.0   
Crispix                       K    C       110        2    0     220    1.0   
Double Chex                   R    C       100        2    0     190    1.0   
Grape Nuts Flakes             P    C       100        3    1     140    3.0   
Grape-Nuts                    P    C       110        3    0     170    3.0   
Great Grains Pecan            P    C       120        3    3      75    3.0   
Just Right Crunchy  Nuggets   K    C       110        2    1     170    1.0   
Kix                           G    C       110        2    1     260    0.0   
Life                          Q    C       100        4    2     150    2.0   
Maypo                         A    H       100        4    1       0    0.0   
Multi-Grain Cheerios          G    C       100        2    1     220    2.0   
Nutri-grain Wheat             K    C        90        3    0     170    3.0   
Product 19                    K    C       100        3    0     320    1.0   
Puffed Wheat                  Q    C        50        2    0       0    1.0   
Quaker Oat Squares            Q    C       100        4    1     135    2.0   
Quaker Oatmeal                Q    H       100        5    2       0    2.7   
Raisin Squares                K    C        90        2    0       0    2.0   
Rice Krispies                 K    C       110        2    0     290    0.0   
Shredded Wheat                N    C        80        2    0       0    3.0   
Shredded Wheat 'n'Bran        N    C        90        3    0       0    4.0   
Shredded Wheat spoon size     N    C        90        3    0       0    3.0   
Special K                     K    C       110        6    0     230    1.0   
Strawberry Fruit Wheats       N    C        90        2    0      15    3.0   
Total Corn Flakes             G    C       110        2    1     200    0.0   
Total Whole Grain             G    C       100        3    1     200    3.0   
Triples                       G    C       110        2    1     250    0.0   
Wheat Chex                    R    C       100        3    1     230    3.0   
Wheaties                      G    C       100        3    1     200    3.0   

                             carbo  sugars  potass  vitamins  shelf  weight  \
name                                                                          
100% Bran                      5.0       6     280        25      3    1.00   
All-Bran                       7.0       5     320        25      3    1.00   
All-Bran with Extra Fiber      8.0       0     330        25      3    1.00   
Bran Chex                     15.0       6     125        25      1    1.00   
Bran Flakes                   13.0       5     190        25      3    1.00   
Cheerios                      17.0       1     105        25      1    1.00   
Corn Chex                     22.0       3      25        25      1    1.00   
Corn Flakes                   21.0       2      35        25      1    1.00   
Cream of Wheat (Quick)        21.0       0      -1         0      2    1.00   
Crispix                       21.0       3      30        25      3    1.00   
Double Chex                   18.0       5      80        25      3    1.00  

### Optional Exercise 1
- Can you find the average sugar content of the cereals which list the portion `cups` size as 1.0?
- Can you find the highest and lowest calorie content of the previous selection?

### Optional Exercise 2
- How many cereals by manufacturer `G` have a higher calorie content than 100?

### Handling duplicates
Next we will be using the `IMDB-Movie-Data.csv` dataset, it contains a list of movies, its genre, a brief description, director, actors, year, runtime, rating, votes, revenue and metascore.

This dataset does not have duplicates initially, but we can artificially create duplicates by appending the dataset to itself.

In [27]:
movies = pd.read_csv("IMDB-Movie-Data.csv", index_col='Title')
movies.head()

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   

                         Revenue (Millions)  Metascore  
Title                                                   
Guardians of the Galaxy              333.13       76.0  
Prometheus                           126.46       65.0  
Split                                138.12       62.0  
Sing                                 270.32       59.0  
Suicide Squad                        325.02       40.0

In [28]:
movies.shape

(1000, 11)

In [29]:
# now lets duplicate the data
# concatenate the dataframe to itself
dup_movies = pd.concat([movies, movies])
dup_movies.head()

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   

                         Revenue (Millions)  Metascore  
Title                                                   
Guardians of the Galaxy              333.13       76.0  
Prometheus                           126.46       65.0  
Split                                138.12       62.0  
Sing                                 270.32       59.0  
Suicide Squad                        325.02       40.0

In [30]:
dup_movies.tail()

Rank                  Genre  \
Title                                                 
Secret in Their Eyes     996    Crime,Drama,Mystery   
Hostel: Part II          997                 Horror   
Step Up 2: The Streets   998    Drama,Music,Romance   
Search Party             999       Adventure,Comedy   
Nine Lives              1000  Comedy,Family,Fantasy   

                                                              Description  \
Title                                                                       
Secret in Their Eyes    A tight-knit team of rising investigators, alo...   
Hostel: Part II         Three American college students studying abroa...   
Step Up 2: The Streets  Romantic sparks occur between two dance studen...   
Search Party            A pair of friends embark on a mission to reuni...   
Nine Lives              A stuffy businessman finds himself trapped ins...   

                                Director  \
Title                                      
Secret in Their Eyes           Billy Ray   
Hostel: Part II                 Eli Roth   
Step Up 2: The Streets        Jon M. Chu   
Search Party              Scot Armstrong   
Nine Lives              Barry Sonnenfeld   

                                                                   Actors  \
Title                                                                       
Secret in Their Eyes    Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II         Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets  Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party            Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives              Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                        Year  Runtime (Minutes)  Rating  Votes  \
Title                                                            
Secret in Their Eyes    2015                111     6.2  27585   
Hostel: Part II         2007                 94     5.5  73152   
Step Up 2: The Streets  2008                 98     6.2  70699   
Search Party            2014                 93     5.6   4881   
Nine Lives              2016                 87     5.3  12435   

                        Revenue (Millions)  Metascore  
Title                                                  
Secret in Their Eyes                   NaN       45.0  
Hostel: Part II                      17.54       46.0  
Step Up 2: The Streets               58.01       50.0  
Search Party                           NaN       22.0  
Nine Lives                           19.64       11.0

We now have 2000 movies, 1000 of which are duplicates

In [31]:
dup_movies.shape

(2000, 11)

We can use the `drop_duplicates(inplace=True)` method to delete the same Dataframe we are working with, this way we don't have to store the DF to a variable

In [32]:
dup_movies.drop_duplicates(inplace=True)

In [33]:
dup_movies.shape

(1000, 11)

The `drop_duplicates()` also has the `keep` argument, which can do the following:
- `first` (default) drop duplicates except for the first occurrence
- `last` drop duplicates except for the last occurrence
- `False` drop all duplicates/

### Cleaning columns
We can change the name of our columns to make them easier to work with, remove symbols, spaces, typos.

Lets print the columns of our original movie dataset.

In [34]:
movies.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')

We can rename columns using a dictionary and the `rename()` method. New names are passed as a dictionary.

In [35]:
movies.rename(columns={'Runtime (Minutes)' : 'Runtime',
                        'Revenue (Millions)' : 'Revenue'}, inplace=True)
movies.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year', 'Runtime',
       'Rating', 'Votes', 'Revenue', 'Metascore'],
      dtype='object')

In [36]:
movies.Revenue

Title
Guardians of the Galaxy    333.13
Prometheus                 126.46
Split                      138.12
Sing                       270.32
Suicide Squad              325.02
                            ...  
Secret in Their Eyes          NaN
Hostel: Part II             17.54
Step Up 2: The Streets      58.01
Search Party                  NaN
Nine Lives                  19.64
Name: Revenue, Length: 1000, dtype: float64

In [37]:
movies.Runtime

Title
Guardians of the Galaxy    121
Prometheus                 124
Split                      117
Sing                       108
Suicide Squad              123
                          ... 
Secret in Their Eyes       111
Hostel: Part II             94
Step Up 2: The Streets      98
Search Party                93
Nine Lives                  87
Name: Runtime, Length: 1000, dtype: int64

We can also rename by using the `columns` property and passing a list of names. Lets have all of our names be lowercase.

In [38]:
movies.columns = ['rank', 'genre', 'description', 'director', 'actors', 'year', 'runtime', 'rating', 'votes', 'revenue_millions', 'metascore']
movies.columns

Index(['rank', 'genre', 'description', 'director', 'actors', 'year', 'runtime',
       'rating', 'votes', 'revenue_millions', 'metascore'],
      dtype='object')

We can also simplify renaning using list comprehension.

In [39]:
movies.columns = [col.upper() for col in movies]
movies.columns

Index(['RANK', 'GENRE', 'DESCRIPTION', 'DIRECTOR', 'ACTORS', 'YEAR', 'RUNTIME',
       'RATING', 'VOTES', 'REVENUE_MILLIONS', 'METASCORE'],
      dtype='object')

In [40]:
movies.columns = [col.lower() for col in movies]
movies.columns

Index(['rank', 'genre', 'description', 'director', 'actors', 'year', 'runtime',
       'rating', 'votes', 'revenue_millions', 'metascore'],
      dtype='object')

### Working with missing values
When working with datasets, we are almost guaranteed to find missing data or null values. We can sometimes find `None` of Numpy's `np.nan`. We have two options:
- Get rid of rows or columns with null values.
- Replace null values with something else.
First, lets check if our movie columns contains null values.

`isnull()` will return if each cell in the dataset has a null value, evaluating to `True` or `False`.

We can then count the number of nulls in each column using the `sum()` method.

In [41]:
movies.isnull()

rank  genre  description  director  actors   year  \
Title                                                                         
Guardians of the Galaxy  False  False        False     False   False  False   
Prometheus               False  False        False     False   False  False   
Split                    False  False        False     False   False  False   
Sing                     False  False        False     False   False  False   
Suicide Squad            False  False        False     False   False  False   
...                        ...    ...          ...       ...     ...    ...   
Secret in Their Eyes     False  False        False     False   False  False   
Hostel: Part II          False  False        False     False   False  False   
Step Up 2: The Streets   False  False        False     False   False  False   
Search Party             False  False        False     False   False  False   
Nine Lives               False  False        False     False   False  False   

                         runtime  rating  votes  revenue_millions  metascore  
Title                                                                         
Guardians of the Galaxy    False   False  False             False      False  
Prometheus                 False   False  False             False      False  
Split                      False   False  False             False      False  
Sing                       False   False  False             False      False  
Suicide Squad              False   False  False             False      False  
...                          ...     ...    ...               ...        ...  
Secret in Their Eyes       False   False  False              True      False  
Hostel: Part II            False   False  False             False      False  
Step Up 2: The Streets     False   False  False             False      False  
Search Party               False   False  False              True      False  
Nine Lives                 False   False  False             False      False  

[1000 rows x 11 columns]

In [42]:
movies.isnull().sum()

rank                  0
genre                 0
description           0
director              0
actors                0
year                  0
runtime               0
rating                0
votes                 0
revenue_millions    128
metascore            64
dtype: int64

We can see that our `revenue_millions` column and `metascore` column have null values. To remove a row containing a null value you can use `.dropna()`. To remove a column containing a null value you can change the axis by running `.dropna(axis=1)`. Remember the first axis we get on the shape of a DataFrame is rows, and the second one is columns, this is where the axis 0 = rows, and axis 1 = columns comes from.

Lets see a dropping example with a temporary DF.

In [43]:
# Copy the original Dataframe into a new variable
drop_movies = movies.copy()
drop_movies.head()

rank                     genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   

                                                               description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   

                                     director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   

                                                                    actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   

                         year  runtime  rating   votes  revenue_millions  \
Title                                                                      
Guardians of the Galaxy  2014      121     8.1  757074            333.13   
Prometheus               2012      124     7.0  485820            126.46   
Split                    2016      117     7.3  157606            138.12   
Sing                     2016      108     7.2   60545            270.32   
Suicide Squad            2016      123     6.2  393727            325.02   

                         metascore  
Title                               
Guardians of the Galaxy       76.0  
Prometheus                    65.0  
Split                         62.0  
Sing                          59.0  
Suicide Squad                 40.0

In [44]:
drop_movies.shape

(1000, 11)

In [45]:
drop_movies.isnull().sum()

rank                  0
genre                 0
description           0
director              0
actors                0
year                  0
runtime               0
rating                0
votes                 0
revenue_millions    128
metascore            64
dtype: int64

In [46]:
drop_movies.dropna(inplace=True)

In [47]:
drop_movies.shape

(838, 11)

In [48]:
drop_movies.isnull().sum()

rank                0
genre               0
description         0
director            0
actors              0
year                0
runtime             0
rating              0
votes               0
revenue_millions    0
metascore           0
dtype: int64

It seems like a waste to drop all of those rows that were missing some data. Lets replace the missing values with something else in our original dataframe `movies`

In [49]:
movies.isnull().sum()

rank                  0
genre                 0
description           0
director              0
actors                0
year                  0
runtime               0
rating                0
votes                 0
revenue_millions    128
metascore            64
dtype: int64

In [50]:
# Select the revenue column
revenue = movies['revenue_millions']

In [51]:
# Series type, not dataframe
revenue

Title
Guardians of the Galaxy    333.13
Prometheus                 126.46
Split                      138.12
Sing                       270.32
Suicide Squad              325.02
                            ...  
Secret in Their Eyes          NaN
Hostel: Part II             17.54
Step Up 2: The Streets      58.01
Search Party                  NaN
Nine Lives                  19.64
Name: revenue_millions, Length: 1000, dtype: float64

In [52]:
# Calculate revenue mean
revenue_mean = revenue.mean()
revenue_mean

82.95637614678898

Lets fill up our missing values with our new `revenue_mean` value using the `.fillna()` method. Remember to use the `inplace=True` to affect the original **Dataframe**.

In [53]:
revenue.fillna(revenue_mean, inplace=True)

In [54]:
movies.isnull().sum()

rank                 0
genre                0
description          0
director             0
actors               0
year                 0
runtime              0
rating               0
votes                0
revenue_millions     0
metascore           64
dtype: int64

### Optional exercise
- Fill out the metascore missing values with the mean of metascores.
- Verify that the `movies` DF has no null values.

### Understanding your variables
Remember you can get a summary of your data using the `describe()` method. Lets describe `movies`

In [55]:
movies.describe()

rank         year      runtime       rating         votes  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1.000000e+03   
mean    500.500000  2012.783000   113.172000     6.723200  1.698083e+05   
std     288.819436     3.205962    18.810908     0.945429  1.887626e+05   
min       1.000000  2006.000000    66.000000     1.900000  6.100000e+01   
25%     250.750000  2010.000000   100.000000     6.200000  3.630900e+04   
50%     500.500000  2014.000000   111.000000     6.800000  1.107990e+05   
75%     750.250000  2016.000000   123.000000     7.400000  2.399098e+05   
max    1000.000000  2016.000000   191.000000     9.000000  1.791916e+06   

       revenue_millions   metascore  
count       1000.000000  936.000000  
mean          82.956376   58.985043  
std           96.412043   17.194757  
min            0.000000   11.000000  
25%           17.442500   47.000000  
50%           60.375000   59.500000  
75%           99.177500   72.000000  
max          936.630000  100.000000

To describe a single column you can select the column and run the method. You can also describe string columns, like `genre`

In [56]:
movies.genre.describe()

count                        1000
unique                        207
top       Action,Adventure,Sci-Fi
freq                           50
Name: genre, dtype: object

To count frequency of data you can use `value_counts()`

In [57]:
# Show top 10 genres
movies.genre.value_counts().head(10)

Action,Adventure,Sci-Fi       50
Drama                         48
Comedy,Drama,Romance          35
Comedy                        32
Drama,Romance                 31
Animation,Adventure,Comedy    27
Action,Adventure,Fantasy      27
Comedy,Drama                  27
Comedy,Romance                26
Crime,Drama,Thriller          24
Name: genre, dtype: int64

We can further try to describe relationship in our variables by using the `.corr()` method, this will show the relationship between two variables (bivariate relationship). Positive numbers indicate a positive correlation, one goes up when the other variable goes up. Negative numbers indicate an inverse correlation - one goes up the other goes down. 1.0 indicates a perfect correlation.

In [58]:
movies.corr()

rank      year   runtime    rating     votes  \
rank              1.000000 -0.261605 -0.221739 -0.219555 -0.283876   
year             -0.261605  1.000000 -0.164900 -0.211219 -0.411904   
runtime          -0.221739 -0.164900  1.000000  0.392214  0.407062   
rating           -0.219555 -0.211219  0.392214  1.000000  0.511537   
votes            -0.283876 -0.411904  0.407062  0.511537  1.000000   
revenue_millions -0.252996 -0.117562  0.247834  0.189527  0.607941   
metascore        -0.191869 -0.079305  0.211978  0.631897  0.325684   

                  revenue_millions  metascore  
rank                     -0.252996  -0.191869  
year                     -0.117562  -0.079305  
runtime                   0.247834   0.211978  
rating                    0.189527   0.631897  
votes                     0.607941   0.325684  
revenue_millions          1.000000   0.133328  
metascore                 0.133328   1.000000

### Manipulating Dataframes
On the next section we will further explore how to explore Dataframe and Series:
- Selecting columns
- Selecting rows
- Conditional selections

By now we now how to select columns using square brackets

In [60]:
genre_col = movies['genre']
genre_col

Title
Guardians of the Galaxy     Action,Adventure,Sci-Fi
Prometheus                 Adventure,Mystery,Sci-Fi
Split                               Horror,Thriller
Sing                        Animation,Comedy,Family
Suicide Squad              Action,Adventure,Fantasy
                                     ...           
Secret in Their Eyes            Crime,Drama,Mystery
Hostel: Part II                              Horror
Step Up 2: The Streets          Drama,Music,Romance
Search Party                       Adventure,Comedy
Nine Lives                    Comedy,Family,Fantasy
Name: genre, Length: 1000, dtype: object

The following code returns a Series. To extract a column as a DataFrame you need to pass a list of column names, in our case a single column

In [61]:
type(genre_col)

pandas.core.series.Series

In [62]:
genre_col = movies[['genre']]
type(genre_col)

pandas.core.frame.DataFrame

Since its just a list, adding another column name is easy

In [63]:
subset = movies[['genre', 'rating']]
subset.head()

genre  rating
Title                                                    
Guardians of the Galaxy   Action,Adventure,Sci-Fi     8.1
Prometheus               Adventure,Mystery,Sci-Fi     7.0
Split                             Horror,Thriller     7.3
Sing                      Animation,Comedy,Family     7.2
Suicide Squad            Action,Adventure,Fantasy     6.2

#### Data by rows
For rows we have two options:
- `.loc` - locates by name
- `.iloc` - locates by numerical index

We are still indexed by movie Title so to use `.loc` we give the Title of the movie

In [64]:
prom = movies.loc['Prometheus']
prom

rank                                                                2
genre                                        Adventure,Mystery,Sci-Fi
description         Following clues to the origin of mankind, a te...
director                                                 Ridley Scott
actors              Noomi Rapace, Logan Marshall-Green, Michael Fa...
year                                                             2012
runtime                                                           124
rating                                                            7.0
votes                                                          485820
revenue_millions                                               126.46
metascore                                                        65.0
Name: Prometheus, dtype: object

With `.iloc` we would give the numerical index of the movie

In [66]:
movies.iloc[1]

rank                                                                2
genre                                        Adventure,Mystery,Sci-Fi
description         Following clues to the origin of mankind, a te...
director                                                 Ridley Scott
actors              Noomi Rapace, Logan Marshall-Green, Michael Fa...
year                                                             2012
runtime                                                           124
rating                                                            7.0
votes                                                          485820
revenue_millions                                               126.46
metascore                                                        65.0
Name: Prometheus, dtype: object

These methods can be thought of as similar to `list` slicing. Lets select multiple rows.

In [68]:
subset = movies.loc['Prometheus':'Sing']
subset

rank                     genre  \
Title                                        
Prometheus     2  Adventure,Mystery,Sci-Fi   
Split          3           Horror,Thriller   
Sing           4   Animation,Comedy,Family   

                                                  description  \
Title                                                           
Prometheus  Following clues to the origin of mankind, a te...   
Split       Three girls are kidnapped by a man with a diag...   
Sing        In a city of humanoid animals, a hustling thea...   

                        director  \
Title                              
Prometheus          Ridley Scott   
Split         M. Night Shyamalan   
Sing        Christophe Lourdelet   

                                                       actors  year  runtime  \
Title                                                                          
Prometheus  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012      124   
Split       James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016      117   
Sing        Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016      108   

            rating   votes  revenue_millions  metascore  
Title                                                    
Prometheus     7.0  485820            126.46       65.0  
Split          7.3  157606            138.12       62.0  
Sing           7.2   60545            270.32       59.0

In [69]:
# Or using iloc
movies.iloc[1:4]

rank                     genre  \
Title                                        
Prometheus     2  Adventure,Mystery,Sci-Fi   
Split          3           Horror,Thriller   
Sing           4   Animation,Comedy,Family   

                                                  description  \
Title                                                           
Prometheus  Following clues to the origin of mankind, a te...   
Split       Three girls are kidnapped by a man with a diag...   
Sing        In a city of humanoid animals, a hustling thea...   

                        director  \
Title                              
Prometheus          Ridley Scott   
Split         M. Night Shyamalan   
Sing        Christophe Lourdelet   

                                                       actors  year  runtime  \
Title                                                                          
Prometheus  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012      124   
Split       James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016      117   
Sing        Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016      108   

            rating   votes  revenue_millions  metascore  
Title                                                    
Prometheus     7.0  485820            126.46       65.0  
Split          7.3  157606            138.12       62.0  
Sing           7.2   60545            270.32       59.0

#### Conditional selection
What if we want to make a conditional selection? Lets say we want to filter our movies to show only films directed by Ridley Scott or films with a rating greater than or equal to 8.0.

Here is an example of those conditions

In [72]:
condition = (movies['director'] == 'Ridley Scott')
condition.head()

Title
Guardians of the Galaxy    False
Prometheus                  True
Split                      False
Sing                       False
Suicide Squad              False
Name: director, dtype: bool

In [75]:
# Filter out movies that dont fulfill the condition
# Show only movies by this director
# Select movies df where movies director equals Ridley Scott
movies[movies['director'] == 'Ridley Scott']

rank                     genre  \
Title                                                    
Prometheus                 2  Adventure,Mystery,Sci-Fi   
The Martian              103    Adventure,Drama,Sci-Fi   
Robin Hood               388    Action,Adventure,Drama   
American Gangster        471     Biography,Crime,Drama   
Exodus: Gods and Kings   517    Action,Adventure,Drama   
The Counselor            522      Crime,Drama,Thriller   
A Good Year              531      Comedy,Drama,Romance   
Body of Lies             738      Action,Drama,Romance   

                                                              description  \
Title                                                                       
Prometheus              Following clues to the origin of mankind, a te...   
The Martian             An astronaut becomes stranded on Mars after hi...   
Robin Hood              In 12th century England, Robin and his band of...   
American Gangster       In 1970s America, a detective works to bring d...   
Exodus: Gods and Kings  The defiant leader Moses rises up against the ...   
The Counselor           A lawyer finds himself in over his head when h...   
A Good Year             A British investment broker inherits his uncle...   
Body of Lies            A CIA agent on the ground in Jordan hunts down...   

                            director  \
Title                                  
Prometheus              Ridley Scott   
The Martian             Ridley Scott   
Robin Hood              Ridley Scott   
American Gangster       Ridley Scott   
Exodus: Gods and Kings  Ridley Scott   
The Counselor           Ridley Scott   
A Good Year             Ridley Scott   
Body of Lies            Ridley Scott   

                                                                   actors  \
Title                                                                       
Prometheus              Noomi Rapace, Logan Marshall-Green, Michael Fa...   
The Martian             Matt Damon, Jessica Chastain, Kristen Wiig, Ka...   
Robin Hood              Russell Crowe, Cate Blanchett, Matthew Macfady...   
American Gangster       Denzel Washington, Russell Crowe, Chiwetel Eji...   
Exodus: Gods and Kings  Christian Bale, Joel Edgerton, Ben Kingsley, S...   
The Counselor           Michael Fassbender, Penélope Cruz, Cameron Dia...   
A Good Year             Russell Crowe, Abbie Cornish, Albert Finney, M...   
Body of Lies            Leonardo DiCaprio, Russell Crowe, Mark Strong,...   

                        year  runtime  rating   votes  revenue_millions  \
Title                                                                     
Prometheus              2012      124     7.0  485820            126.46   
The Martian             2015      144     8.0  556097            228.43   
Robin Hood              2010      140     6.7  221117            105.22   
American Gangster       2007      157     7.8  337835            130.13   
Exodus: Gods and Kings  2014      150     6.0  137299             65.01   
The Counselor           2013      117     5.3   84927             16.97   
A Good Year             2006      117     6.9   74674              7.46   
Body of Lies            2008      128     7.1  182305             39.38   

                        metascore  
Title                              
Prometheus                   65.0  
The Martian                  80.0  
Robin Hood                   53.0  
American Gangster            76.0  
Exodus: Gods and Kings       52.0  
The Counselor                48.0  
A Good Year                  47.0  
Body of Lies                 57.0

We can create more conditions by using the logical operators `|` "or" and `&` "and".

Lets filter to show movies only by Christopher Nolan or Ridley Scott

In [76]:
movies[(movies['director'] == 'Christopher Nolan') | (movies['director'] == 'Ridley Scott')]

rank                     genre  \
Title                                                    
Prometheus                 2  Adventure,Mystery,Sci-Fi   
Interstellar              37    Adventure,Drama,Sci-Fi   
The Dark Knight           55        Action,Crime,Drama   
The Prestige              65      Drama,Mystery,Sci-Fi   
Inception                 81   Action,Adventure,Sci-Fi   
The Martian              103    Adventure,Drama,Sci-Fi   
The Dark Knight Rises    125           Action,Thriller   
Robin Hood               388    Action,Adventure,Drama   
American Gangster        471     Biography,Crime,Drama   
Exodus: Gods and Kings   517    Action,Adventure,Drama   
The Counselor            522      Crime,Drama,Thriller   
A Good Year              531      Comedy,Drama,Romance   
Body of Lies             738      Action,Drama,Romance   

                                                              description  \
Title                                                                       
Prometheus              Following clues to the origin of mankind, a te...   
Interstellar            A team of explorers travel through a wormhole ...   
The Dark Knight         When the menace known as the Joker wreaks havo...   
The Prestige            Two stage magicians engage in competitive one-...   
Inception               A thief, who steals corporate secrets through ...   
The Martian             An astronaut becomes stranded on Mars after hi...   
The Dark Knight Rises   Eight years after the Joker's reign of anarchy...   
Robin Hood              In 12th century England, Robin and his band of...   
American Gangster       In 1970s America, a detective works to bring d...   
Exodus: Gods and Kings  The defiant leader Moses rises up against the ...   
The Counselor           A lawyer finds himself in over his head when h...   
A Good Year             A British investment broker inherits his uncle...   
Body of Lies            A CIA agent on the ground in Jordan hunts down...   

                                 director  \
Title                                       
Prometheus                   Ridley Scott   
Interstellar            Christopher Nolan   
The Dark Knight         Christopher Nolan   
The Prestige            Christopher Nolan   
Inception               Christopher Nolan   
The Martian                  Ridley Scott   
The Dark Knight Rises   Christopher Nolan   
Robin Hood                   Ridley Scott   
American Gangster            Ridley Scott   
Exodus: Gods and Kings       Ridley Scott   
The Counselor                Ridley Scott   
A Good Year                  Ridley Scott   
Body of Lies                 Ridley Scott   

                                                                   actors  \
Title                                                                       
Prometheus              Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Interstellar            Matthew McConaughey, Anne Hathaway, Jessica Ch...   
The Dark Knight         Christian Bale, Heath Ledger, Aaron Eckhart,Mi...   
The Prestige            Christian Bale, Hugh Jackman, Scarlett Johanss...   
Inception               Leonardo DiCaprio, Joseph Gordon-Levitt, Ellen...   
The Martian             Matt Damon, Jessica Chastain, Kristen Wiig, Ka...   
The Dark Knight Rises   Christian Bale, Tom Hardy, Anne Hathaway,Gary ...   
Robin Hood              Russell Crowe, Cate Blanchett, Matthew Macfady...   
American Gangster       Denzel Washington, Russell Crowe, Chiwetel Eji...   
Exodus: Gods and Kings  Christian Bale, Joel Edgerton, Ben Kingsley, S...   
The Counselor           Michael Fassbender, Penélope Cruz, Cameron Dia...   
A Good Year             Russell Crowe, Abbie Cornish, Albert Finney, M...   
Body of Lies            Leonardo DiCaprio, Russell Crowe, Mark Strong,...   

                        year  runtime  rating    votes  revenue_millions  \
Title                                                                      
Prometheus   

We could also use the `isin()` method to make this more concise

In [78]:
movies[movies['director'].isin(['Christopher Nolan', 'Ridley Scott'])]

rank                     genre  \
Title                                                    
Prometheus                 2  Adventure,Mystery,Sci-Fi   
Interstellar              37    Adventure,Drama,Sci-Fi   
The Dark Knight           55        Action,Crime,Drama   
The Prestige              65      Drama,Mystery,Sci-Fi   
Inception                 81   Action,Adventure,Sci-Fi   
The Martian              103    Adventure,Drama,Sci-Fi   
The Dark Knight Rises    125           Action,Thriller   
Robin Hood               388    Action,Adventure,Drama   
American Gangster        471     Biography,Crime,Drama   
Exodus: Gods and Kings   517    Action,Adventure,Drama   
The Counselor            522      Crime,Drama,Thriller   
A Good Year              531      Comedy,Drama,Romance   
Body of Lies             738      Action,Drama,Romance   

                                                              description  \
Title                                                                       
Prometheus              Following clues to the origin of mankind, a te...   
Interstellar            A team of explorers travel through a wormhole ...   
The Dark Knight         When the menace known as the Joker wreaks havo...   
The Prestige            Two stage magicians engage in competitive one-...   
Inception               A thief, who steals corporate secrets through ...   
The Martian             An astronaut becomes stranded on Mars after hi...   
The Dark Knight Rises   Eight years after the Joker's reign of anarchy...   
Robin Hood              In 12th century England, Robin and his band of...   
American Gangster       In 1970s America, a detective works to bring d...   
Exodus: Gods and Kings  The defiant leader Moses rises up against the ...   
The Counselor           A lawyer finds himself in over his head when h...   
A Good Year             A British investment broker inherits his uncle...   
Body of Lies            A CIA agent on the ground in Jordan hunts down...   

                                 director  \
Title                                       
Prometheus                   Ridley Scott   
Interstellar            Christopher Nolan   
The Dark Knight         Christopher Nolan   
The Prestige            Christopher Nolan   
Inception               Christopher Nolan   
The Martian                  Ridley Scott   
The Dark Knight Rises   Christopher Nolan   
Robin Hood                   Ridley Scott   
American Gangster            Ridley Scott   
Exodus: Gods and Kings       Ridley Scott   
The Counselor                Ridley Scott   
A Good Year                  Ridley Scott   
Body of Lies                 Ridley Scott   

                                                                   actors  \
Title                                                                       
Prometheus              Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Interstellar            Matthew McConaughey, Anne Hathaway, Jessica Ch...   
The Dark Knight         Christian Bale, Heath Ledger, Aaron Eckhart,Mi...   
The Prestige            Christian Bale, Hugh Jackman, Scarlett Johanss...   
Inception               Leonardo DiCaprio, Joseph Gordon-Levitt, Ellen...   
The Martian             Matt Damon, Jessica Chastain, Kristen Wiig, Ka...   
The Dark Knight Rises   Christian Bale, Tom Hardy, Anne Hathaway,Gary ...   
Robin Hood              Russell Crowe, Cate Blanchett, Matthew Macfady...   
American Gangster       Denzel Washington, Russell Crowe, Chiwetel Eji...   
Exodus: Gods and Kings  Christian Bale, Joel Edgerton, Ben Kingsley, S...   
The Counselor           Michael Fassbender, Penélope Cruz, Cameron Dia...   
A Good Year             Russell Crowe, Abbie Cornish, Albert Finney, M...   
Body of Lies            Leonardo DiCaprio, Russell Crowe, Mark Strong,...   

                        year  runtime  rating    votes  revenue_millions  \
Title                                                                      
Prometheus   

Can you get all movies with the following conditions?
- Released between 2005 and 2010
- Have a rating above 8.0
- Made below the 25th percentile in revenue

In [79]:
movies[((movies['year'] >= 2005) & (movies['year'] <= 2010)) # get movies between years 2005 and 2010
      & (movies['rating'] > 8.0) # get movies with rating greater than 8.0
      & (movies['revenue_millions'] < movies['revenue_millions'].quantile(0.25))] # get movies with revenue less than 25th percentile

rank               genre  \
Title                                           
3 Idiots              431        Comedy,Drama   
The Lives of Others   477      Drama,Thriller   
Incendies             714   Drama,Mystery,War   
Taare Zameen Par      992  Drama,Family,Music   

                                                           description  \
Title                                                                    
3 Idiots             Two friends are searching for their long lost ...   
The Lives of Others  In 1984 East Berlin, an agent of the secret po...   
Incendies            Twins journey to the Middle East to discover t...   
Taare Zameen Par     An eight-year-old boy is thought to be a lazy ...   

                                             director  \
Title                                                   
3 Idiots                              Rajkumar Hirani   
The Lives of Others  Florian Henckel von Donnersmarck   
Incendies                            Denis Villeneuve   
Taare Zameen Par                           Aamir Khan   

                                                                actors  year  \
Title                                                                          
3 Idiots               Aamir Khan, Madhavan, Mona Singh, Sharman Joshi  2009   
The Lives of Others  Ulrich Mühe, Martina Gedeck,Sebastian Koch, Ul...  2006   
Incendies            Lubna Azabal, Mélissa Désormeaux-Poulin, Maxim...  2010   
Taare Zameen Par     Darsheel Safary, Aamir Khan, Tanay Chheda, Sac...  2007   

                     runtime  rating   votes  revenue_millions  metascore  
Title                                                                      
3 Idiots                 170     8.4  238789              6.52       67.0  
The Lives of Others      137     8.5  278103             11.28       89.0  
Incendies                131     8.2   92863              6.86       80.0  
Taare Zameen Par         165     8.5  102697              1.20       42.0

If you want to check this, you can remember we used `.describe()` to get the 25th percentile of revenue and it was about 17.4, we can access the value directly by using the `quantile()` method with a float of 0.25.

In [80]:
movies.describe()

rank         year      runtime       rating         votes  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1.000000e+03   
mean    500.500000  2012.783000   113.172000     6.723200  1.698083e+05   
std     288.819436     3.205962    18.810908     0.945429  1.887626e+05   
min       1.000000  2006.000000    66.000000     1.900000  6.100000e+01   
25%     250.750000  2010.000000   100.000000     6.200000  3.630900e+04   
50%     500.500000  2014.000000   111.000000     6.800000  1.107990e+05   
75%     750.250000  2016.000000   123.000000     7.400000  2.399098e+05   
max    1000.000000  2016.000000   191.000000     9.000000  1.791916e+06   

       revenue_millions   metascore  
count       1000.000000  936.000000  
mean          82.956376   58.985043  
std           96.412043   17.194757  
min            0.000000   11.000000  
25%           17.442500   47.000000  
50%           60.375000   59.500000  
75%           99.177500   72.000000  
max          936.630000  100.000000

### Finally a brief look at functions
We will cover functions more in detail in a future lesson, for now you should know that iterating over a DataFrame as it was a list, for example with a for loop, is very slow. An efficient alternative to this is using the `apply()` function. This method passes every value in the DataFrame through a function using vectorization.

Vectorization is a style of programming where operations are applied to whole arrays instead of individual elements, this makes it really efficient.

Lets create a function to rate our movies. The following code creates a function that will return "good" if the rating is equals or greater than 8.0, else it will return false. This rating will be created as a new column in our DataFrame.

In [81]:
# Create our function
# Will return "good" if x is greater or equals to 8.0
# Will return "bad" if x is less than 8.0

# def is a keyword to define functions. We will see more of this in a future lesson
def rating_function(x):
    if x >= 8.0:
        return "good"
    else:
        return "bad"

In [82]:
movies['rating_category'] = movies['rating'].apply(rating_function)
movies.head()

rank                     genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   

                                                               description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   

                                     director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   

                                                                    actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   

                         year  runtime  rating   votes  revenue_millions  \
Title                                                                      
Guardians of the Galaxy  2014      121     8.1  757074            333.13   
Prometheus               2012      124     7.0  485820            126.46   
Split                    2016      117     7.3  157606            138.12   
Sing                     2016      108     7.2   60545            270.32   
Suicide Squad            2016      123     6.2  393727            325.02   

                         metascore rating_category  
Title                                               
Guardians of the Galaxy       76.0            good  
Prometheus                    65.0             bad  
Split                         62.0             bad  
Sing                          59.0             bad  
Suicide Squad                 40.0             bad

In [84]:
subset.to_csv('subset.csv') # this is how you save a selection to csv
# the code above will create a new csv called 'subset.csv'

### Optional Exercise
- Select the movies with ratings between 6.0 and 8.5 released between 2010 and 2016. Save this selection to a variable.
- Select movies with runtime greater than 120 and rating greater or equal to 8.0. Save this selection to a variable.
- Save these two selections to two `.csv` files using the `df.to_csv('name_of_csv.csv')` function. Where 'df' is the name of your DataFrame
